File: 2015/Rust/src/bin/02.rs
aoc_2015::main!();


fn generator(input: &str) -> Vec<(u32, u32, u32)> {
    input
        .lines()
        .map(|l| {
            l.split('x')
                .map(|n| n.parse::<u32>().unwrap())
                .collect_tuple::<(_, _, _)>()
                .unwrap()
        })
        .collect()
}

fn part_1(gifts : Vec<(u32, u32, u32)>) -> u32 {
    gifts
        .iter()
        .fold( 0, |acc , g| 
            acc + 2 * (g.0 * g.1 + g.0 * g.2 + g.1 * g.2) + [g.0*g.1, g.0*g.2, g.1*g.2].iter().min().unwrap())
}

fn part_2(gifts : Vec<(u32, u32, u32)>)  -> u32{
    gifts
    .iter()
    .fold(0,  |acc, (l, w, h)| acc + l * w * h + 2 * [l+w,l+h,h+w].iter().min().unwrap())
}

File: 2015/Python/02.py
def generator(input) :
    return [sorted([int(size) for size in present.split('x')]) for present in input.splitlines()]

def part_1(dimensions) :
    return sum([2 * h * (l + w) + 3 * l * w for [l,w,h] in dimensions])

def part_2(dimensions) : 
    return sum([2 * (l + w) + l * w * h for [l,w,h] in dimensions])
